# Name: Angel Vasquez
# Lab Partners: George Rodruiguez, Swati


DataSet: http://kaggle.com/datasets/tejashvi14/travel-insurance-prediction-data



In [ ]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
from sklearn.tree import DecisionTreeClassifier
from sklearn.preprocessing import LabelEncoder
from sklearn.ensemble import RandomForestClassifier
from sklearn.model_selection import train_test_split
from sklearn import metrics

In [ ]:
# load dataset
df = pd.read_csv("TravelInsurancePrediction.csv")

le = LabelEncoder()

cat_cols = ['Employment Type', 'GraduateOrNot', 'FrequentFlyer', 'EverTravelledAbroad']
for col in cat_cols:
    df[col] = le.fit_transform(df[col])

df.head()

,Unnamed: 0,Age,Employment Type,GraduateOrNot,AnnualIncome,FamilyMembers,ChronicDiseases,FrequentFlyer,EverTravelledAbroad,TravelInsurance
0,0,31,0,1,400000,6,1,0,0,0
1,1,31,1,1,1250000,7,0,0,0,0
2,2,34,1,1,500000,4,1,0,0,1
3,3,28,1,1,700000,3,1,0,0,0
4,4,28,1,1,700000,8,1,1,0,0


In [ ]:
#split dataset in features and target variable
x_data=df[['Age','Employment Type','GraduateOrNot','AnnualIncome','FamilyMembers','ChronicDiseases','FrequentFlyer','EverTravelledAbroad']]
y_data=df['TravelInsurance']

In [ ]:
# Split dataset into training set and test set
X_train, X_test, y_train, y_test = train_test_split(x_data, y_data, test_size=0.3, random_state=1)

In [ ]:
# Train a single Decision Tree
dt_model = DecisionTreeClassifier(random_state=42)
dt_model.fit(X_train,y_train)

# Calculate the accuracy and assign it to dt_accuracy
# Train a single Decision Tree
dt_model = DecisionTreeClassifier(random_state=42)
dt_model.fit(X_train, y_train)

# Calculate the accuracy and assign it to dt_accuracy
dt_accuracy = dt_model.score(X_test, y_test)

print("Decision Tree Test Accuracy:", f"{dt_accuracy * 100:.2f}%")

Decision Tree Test Accuracy: 76.88%


In [ ]:
# Train a Random Forest with random feature selection
rf_model = RandomForestClassifier(n_estimators=5, max_features=2, random_state=42)  # max_features=2 means only 2 features are considered at each split

rf_model.fit(X_train, y_train)


# Calculate the accuracy and assign it to rf_accuracy
rf_accuracy = rf_model.score(X_test, y_test)


# Train a Random Forest with random feature selection
rf_model = RandomForestClassifier(n_estimators=5, max_features=2, random_state=42)  # max_features=2 means only 2 features are considered at each split


rf_model.fit(X_train, y_train)


# Calculate the accuracy and assign it to rf_accuracy
rf_accuracy = rf_model.score(X_test, y_test)

# Now you can use rf_accuracy in the print statements below
print("RF Model Accuracy:", f"{rf_accuracy * 100:.2f}", "%")

RF Model Accuracy: 77.89 %


In [ ]:
# Train a Random Forest with random feature selection
rf_model = RandomForestClassifier(n_estimators=5, max_features=2, random_state=42)  # max_features=2 means only 2 features are considered at each split

rf_model.fit(X_train, y_train)

# Make predictions on the test set
y_pred_rf = rf_model.predict(X_test)

# Calculate precision, recall, and F1-score for Random Forest
precision_rf = metrics.precision_score(y_test, y_pred_rf)
recall_rf = metrics.recall_score(y_test, y_pred_rf)
# Train a Random Forest with random feature selection
rf_model = RandomForestClassifier(n_estimators=5, max_features=2, random_state=42)  # max_features=2 means only 2 features are considered at each split

rf_model.fit(X_train, y_train)

# Make predictions on the test set
y_pred_rf = rf_model.predict(X_test)

# Calculate precision, recall, and F1-score for Random Forest
precision_rf = metrics.precision_score(y_test, y_pred_rf)
recall_rf = metrics.recall_score(y_test, y_pred_rf)
f1_rf = metrics.f1_score(y_test, y_pred_rf)


print("RF Precision:", precision_rf)
print("RF Recall:", recall_rf)
print("RF F1-Score:", f1_rf)

RF Precision: 0.7204301075268817
RF Recall: 0.6261682242990654
RF F1-Score: 0.67


In [ ]:
# Prepare a provision for a new data entry and its prediction of an outcome.

def predict_travel_insurance():
    while True:
        try:
            age = int(input("Enter age: "))
            employment_type = int(input("Enter employment type (0 = Government Sector, 1 = Private Sector/Self Employed): "))
            graduate_or_not = int(input("Enter graduate status (0 = No, 1 = Yes): "))
            annual_income = int(input("Enter annual income: "))
            family_members = int(input("Enter number of family members: "))
            chronic_diseases = int(input("Enter chronic diseases (0 = No, 1 = Yes): "))
            frequent_flyer = int(input("Enter frequent flyer status (0 = No, 1 = Yes): "))
            ever_travelled_abroad = int(input("Enter ever travelled abroad (0 = No, 1 = Yes): "))

            # Input validation (example: check for non-negative values)
            if any(val < 0 for val in [age, employment_type, graduate_or_not, annual_income, family_members, chronic_diseases, frequent_flyer, ever_travelled_abroad]):
                print("Error: Please enter non-negative values for all features.")
                continue  #restart the loop

            new_customer_data = pd.DataFrame([[age, employment_type, graduate_or_not, annual_income, family_members, chronic_diseases, frequent_flyer, ever_travelled_abroad]],
                                           columns=['Age', 'Employment Type', 'GraduateOrNot', 'AnnualIncome', 'FamilyMembers', 'ChronicDiseases', 'FrequentFlyer', 'EverTravelledAbroad'])

            prediction = rf_model.predict(new_customer_data)[0]

            if prediction == 1:
                print("Prediction: Customer is likely to purchase travel insurance.")
            else:
                print("Prediction: Customer is likely not to purchase travel insurance.")

            another_customer = input("Enter data for another customer? (yes/no): ")
            if another_customer.lower() != 'yes':
                break #exit loop
        except ValueError:
            print("Invalid input. Please enter valid numerical values.")

predict_travel_insurance()

Enter age: 21
Enter employment type (0 = Government Sector, 1 = Private Sector/Self Employed): 1
Enter graduate status (0 = No, 1 = Yes): 1
Enter annual income: 15000
Enter number of family members: 7
Enter chronic diseases (0 = No, 1 = Yes): 0
Enter frequent flyer status (0 = No, 1 = Yes): 0
Enter ever travelled abroad (0 = No, 1 = Yes): 0
Prediction: Customer is likely not to purchase travel insurance.
Enter data for another customer? (yes/no): no


In [ ]:
# For the feature importance, show a table of features and its corresponding values

feature_importances = pd.DataFrame({'feature': x_data.columns, 'importance': rf_model.feature_importances_})
feature_importances = feature_importances.sort_values('importance', ascending=False)
feature_importances

,feature,importance
3,AnnualIncome,0.379899
4,FamilyMembers,0.205996
0,Age,0.184500
7,EverTravelledAbroad,0.117762
6,FrequentFlyer,0.043961
1,Employment Type,0.024327
5,ChronicDiseases,0.023385
2,GraduateOrNot,0.020170


# Summary:
This report presents the findings of a machine learning analysis conducted on a dataset of 1,987 customers to predict whether a customer is likely to purchase travel insurance. The analysis used a method called Random Forest, which works by building multiple decision trees and combining their results to produce a more accurate and reliable prediction. Before training the model, four text-based columns, Employment Type, Graduate Status, Frequent Flyer, and Ever Travelled Abroad,  were converted into numerical values using a technique called Label Encoding, since machine learning models can only process numbers. The dataset was then split into a training set (70%) used to teach the model, and a test set (30%) used to evaluate how well it learned. Two models were compared: a single Decision Tree, which achieved an accuracy of 76.88%, and a Random Forest, which performed slightly better at 77.89%. The Random Forest model also recorded a Precision of 0.72, meaning that when it predicted a customer would buy insurance, it was correct 72% of the time. Its Recall was 0.63, meaning it successfully identified 63% of all actual buyers. The F1-Score, which balances both Precision and Recall, came in at 0.67, which is  a solid result for this type of prediction problem. When examining which factors had the greatest influence on the prediction, Annual Income stood out as the most important feature by a wide margin, accounting for nearly 38% of the model's decision-making. This was followed by Number of Family Members at 21% and Age at 18%. Whether a customer had ever travelled abroad contributed 12%, while the remaining factors each played a smaller role. Overall, the Random Forest model demonstrates a reasonable ability to identify customers who are likely to purchase travel insurance, with income, family size, and age being the strongest indicators of that decision.
